File: app/src/main/java/com/quranapp/android/frags/readerindex/BaseFragReaderIndex.java
package com.quranapp.android.frags.readerindex;

import android.content.Context;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.annotation.CallSuper;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.SimpleItemAnimator;

import com.quranapp.android.R;
import com.quranapp.android.activities.ActivityReaderIndexPage;
import com.quranapp.android.components.quran.QuranMeta;
import com.quranapp.android.databinding.FragReaderIndexBinding;
import com.quranapp.android.frags.BaseFragment;
import com.quranapp.android.interfaceUtils.readerIndex.FragReaderIndexCallback;
import com.quranapp.android.utils.thread.runner.RunnableTaskRunner;
import com.quranapp.android.utils.thread.tasks.BaseRunnableTask;
import com.quranapp.android.views.helper.RecyclerView2;

import java.util.concurrent.atomic.AtomicReference;

public abstract class BaseFragReaderIndex extends BaseFragment implements FragReaderIndexCallback {
    protected final Handler mHandler = new Handler(Looper.getMainLooper());
    private final RunnableTaskRunner mTaskRunner = new RunnableTaskRunner(mHandler);
    private final AtomicReference<QuranMeta> mQuranMetaRef = new AtomicReference<>();
    protected FragReaderIndexBinding mBinding;
    private boolean mIsReversed;

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);

        if (context instanceof ActivityReaderIndexPage) {
            ((ActivityReaderIndexPage) context).addToCallbacks(this);
        }
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.frag_reader_index, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mBinding = FragReaderIndexBinding.bind(view);
        // ViewUtils.setBounceOverScrollRV(mBinding.list);

        mBinding.loader.setVisibility(View.VISIBLE);

        QuranMeta.prepareInstance(view.getContext(), quranMeta -> {
            mQuranMetaRef.set(quranMeta);

            mTaskRunner.runAsync(new BaseRunnableTask() {
                @Override
                public void postExecute() {
                    mBinding.loader.setVisibility(View.GONE);
                }

                @Override
                public void runTask() {
                    initList(mBinding.list, view.getContext());
                }

                @Override
                public void onComplete() {}
            });
        });
    }

    @CallSuper
    protected void initList(RecyclerView2 list, Context ctx) {
        RecyclerView.ItemAnimator animator = list.getItemAnimator();
        if (animator instanceof SimpleItemAnimator) {
            ((SimpleItemAnimator) animator).setSupportsChangeAnimations(false);
        }
    }

    @Override
    public void scrollToTop(boolean smooth) {
        if (mBinding != null && mBinding.list.getLayoutManager() != null) {
            if (smooth) {
                mBinding.list.smoothScrollToPosition(0);
            } else {
                mBinding.list.scrollToPosition(0);
            }
        }
    }

    @Override
    public void sort(Context ctx) {
        mTaskRunner.runAsync(new BaseRunnableTask() {
            @Override
            public void preExecute() {
                mBinding.loader.setVisibility(View.VISIBLE);
            }

            @Override
            public void postExecute() {
                mBinding.loader.setVisibility(View.GONE);
            }

            @Override
            public void runTask() {
                resetAdapter(mBinding.list, ctx, !mIsReversed);
            }

            @Override
            public void onComplete() {}
        });
    }

    @CallSuper
    protected void resetAdapter(RecyclerView2 list, Context ctx, boolean reverse) {
        mIsReversed = reverse;
    }

    public QuranMeta getQuranMeta() {
        return mQuranMetaRef.get();
    }
}

File: app/src/main/java/com/quranapp/android/frags/readerindex/BaseFragReaderIndex.kt
package com.quranapp.android.frags.readerindex

import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.CallSuper
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.SimpleItemAnimator
import com.quranapp.android.R
import com.quranapp.android.activities.ActivityReaderIndexPage
import com.quranapp.android.components.quran.QuranMeta
import com.quranapp.android.databinding.FragReaderIndexBinding
import com.quranapp.android.frags.BaseFragment
import com.quranapp.android.interfaceUtils.OnResultReadyCallback
import com.quranapp.android.interfaceUtils.readerIndex.FragReaderIndexCallback
import com.quranapp.android.viewModels.FavChaptersViewModel
import com.quranapp.android.views.helper.RecyclerView2
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.concurrent.atomic.AtomicReference

abstract class BaseFragReaderIndex : BaseFragment(), FragReaderIndexCallback {
    lateinit var favChaptersModel: FavChaptersViewModel
    private val quranMetaRef = AtomicReference<QuranMeta>()
    protected lateinit var binding: FragReaderIndexBinding
    private var isReversed = false

    override fun onAttach(context: Context) {
        super.onAttach(context)

        (context as? ActivityReaderIndexPage)?.addToCallbacks(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        return inflater.inflate(R.layout.frag_reader_index, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        favChaptersModel = ViewModelProvider(requireActivity())[FavChaptersViewModel::class.java]
        favChaptersModel.refreshFavChapters(view.context)
        binding = FragReaderIndexBinding.bind(view)

        // ViewUtils.setBounceOverScrollRV(mBinding.list);
        binding.loader.visibility = View.VISIBLE
        QuranMeta.prepareInstance(view.context, object : OnResultReadyCallback<QuranMeta> {
            override fun onReady(r: QuranMeta) {
                quranMetaRef.set(r)

                binding.loader.visibility = View.VISIBLE

                lifecycleScope.launch(Dispatchers.Main) {
                    initList(binding.list, view.context)

                    binding.loader.visibility = View.GONE
                }
            }
        })
    }

    @CallSuper
    protected open fun initList(list: RecyclerView2, ctx: Context) {
        val animator = list.itemAnimator
        if (animator is SimpleItemAnimator) {
            animator.supportsChangeAnimations = false
        }
    }

    override fun scrollToTop(smooth: Boolean) {
        binding.list.layoutManager?.let {
            if (smooth) binding.list.smoothScrollToPosition(0)
            else binding.list.scrollToPosition(0)
        }
    }

    override fun sort(ctx: Context) {
        binding.loader.visibility = View.VISIBLE
        lifecycleScope.launch {
            resetAdapter(binding.list, ctx, !isReversed)

            withContext(Dispatchers.Main) {
                binding.loader.visibility = View.GONE
            }
        }
    }

    @CallSuper
    protected open fun resetAdapter(list: RecyclerView2, ctx: Context, reverse: Boolean) {
        isReversed = reverse
    }

    val quranMeta: QuranMeta get() = quranMetaRef.get()
}